### ISSUE TITLE
deeplearning - 4/w4

### ISSUE THREAD

**Issue by getElementsByName** [getElementsByName/tmp-doc#9]
# Special applications: Face recognition & Neural style transfer
- Discover how CNNs can be applied to multiple fields, including art generation and face recognition. Implement your own algorithm to generate art and recognize faces!
- Face Recognition
   - https://www.youtube.com/watch?v=wr4rx0Spihs
- Neural style transfer
   - https://www.youtube.com/watch?v=aQcj13gX73Y

### What is face recognition?
- Face Verification와 Face Recognition 비교

#### Face Verification 
- input 이미지가 특정 사용자인지 아닌지 판단
- 예) 주어진 이미지가 유재석인가?

#### Face Recognition
- input 이미지가 누구인가?
   - 누구를 판단하기 위한 데이터베이스가 전제
- 예) 주어진 이미지가 누구인가?

### One Shot Learning
- For most face recognition applications you need to be able to recognize a person given just one single image. 
   - wiki: https://en.wikipedia.org/wiki/One-shot_learning
![image](https://user-images.githubusercontent.com/21363169/91301364-c2ca9100-e7df-11ea-86bd-4a94fa91f5f6.png)
- solution: similarity
   - This allows us to learn to recognize a new person given just a single image of that person(one-shot learning problem).
![image](https://user-images.githubusercontent.com/21363169/91301397-d1b14380-e7df-11ea-9847-e3d667359a08.png)

### Siamese Network
![image](https://user-images.githubusercontent.com/21363169/91302223-19849a80-e7e1-11ea-8127-e0c61baa2343.png)
- deep layer에 feature map, embedding 비교
![image](https://user-images.githubusercontent.com/21363169/91302180-07a2f780-e7e1-11ea-8f3f-a4cb7e1c1ad5.png)

> - Siamese twin: https://en.wikipedia.org/wiki/Chang_and_Eng_Bunker
> - Siamese connection: https://en.wikipedia.org/wiki/Siamese_connection

###  Triplet Loss
- Anchor, Positive, Negative 를 하나의 세트로 구성
   - Positive: Anchor와 동일 인물
   - Negative: Anchor와 다른 인물
![image](https://user-images.githubusercontent.com/21363169/91302916-281f8180-e7e2-11ea-8f91-5339fd88f691.png)
- max(..., 0): margin 만큼 차이가 발생하면 이후의 차이는 고려하지 않음
![image](https://user-images.githubusercontent.com/21363169/91302928-2e156280-e7e2-11ea-8147-6024f4ff5266.png)
-  (the idea of a blank net or deep blank is a very popular way of naming algorithms in the deep learning world)
![image](https://user-images.githubusercontent.com/21363169/91302938-31105300-e7e2-11ea-9d72-cef9466914de.png)
![image](https://user-images.githubusercontent.com/21363169/91302959-3a99bb00-e7e2-11ea-906e-f41e3eb59582.png)

### Face Verification and Binary Classification
-  So, this is a way to treat face recognition just as a binary classification problem. And this is an alternative to the triplet loss for training a system like this.
- 유사도를 측정하는 여러가지 variation이 존재
   - 예) Chi-square similarity
![image](https://user-images.githubusercontent.com/21363169/91303626-4174fd80-e7e3-11ea-8586-050e1c75b566.png)
![Screenshot from 2020-08-27 11-33-53](https://user-images.githubusercontent.com/21363169/91377189-cef33480-e859-11ea-9b3d-39f164ed5b1d.png)
![image](https://user-images.githubusercontent.com/21363169/91304234-23f46380-e7e4-11ea-86d6-421a5b0d583d.png)

### What is neural style transfer?
![image](https://user-images.githubusercontent.com/21363169/91310027-54400000-e7ec-11ea-8e83-8180ca4c1172.png)

### What are deep ConvNets learning?
- These are the nine patches that cause one hidden unit to be highly activated. And then each grouping, this is a different set of nine image patches that cause one hidden unit to be activated. 
   - train hidden units in layer 1, they're often looking for relatively simple features such as edge or a particular shade of color. 
   - later units are actually seen larger image patches
![image](https://user-images.githubusercontent.com/21363169/91310336-b6006a00-e7ec-11ea-8f62-bbd1c03a0b75.png)
> - Visualizing what ConvNets learn: https://cs231n.github.io/understanding-cnn/

### Cost Function
![image](https://user-images.githubusercontent.com/21363169/91310988-8140e280-e7ed-11ea-8074-b3cd25b2ccbc.png)
- 새로 생성하는 input 이미지가 학습하는 파라매터
![image](https://user-images.githubusercontent.com/21363169/91311020-8aca4a80-e7ed-11ea-8d70-7448188c06fd.png)

### Content Cost Function
- 중간 hidden layer의 activation 값 차이
![image](https://user-images.githubusercontent.com/21363169/91311138-b2b9ae00-e7ed-11ea-943b-a7f581ced410.png)

### Style Cost Function
- 동일 layer의 필터간 상관관계
![image](https://user-images.githubusercontent.com/21363169/91311287-e1d01f80-e7ed-11ea-97df-526e7b78fc36.png)
![image](https://user-images.githubusercontent.com/21363169/91311345-f2809580-e7ed-11ea-82af-087374da72a5.png)
- Gram Matrix (Style Matrix)
- correlation 으로 표현하는 부분은 수학적으로 엄밀히 unnormalized cross covariance 임. 직관적인 표현을 위해 correlation으로 표현
   - 평균을 빼는 항이 없기 때문
![image](https://user-images.githubusercontent.com/21363169/91311375-fca29400-e7ed-11ea-9e04-2febfd697bcc.png)
![Screenshot from 2020-08-27 11-38-41](https://user-images.githubusercontent.com/21363169/91377275-ff3ad300-e859-11ea-84fa-550b8f00c995.png)

- 각 layer의 Style Matrix 간 차이의 Frobenius norm을 구하고 모두 합한다.
![image](https://user-images.githubusercontent.com/21363169/91311754-67ec6600-e7ee-11ea-860b-b71b8776d859.png)
![Screenshot from 2020-08-27 11-38-44](https://user-images.githubusercontent.com/21363169/91377279-019d2d00-e85a-11ea-830d-b1109ba29021.png)

### 1D and 3D Generalizations
- 2D 데이터 뿐만 아니라 1D/3D 데이터도 CNN을 적용할 수 있음
- 데이터 차원에 대한 Generalizations
![image](https://user-images.githubusercontent.com/21363169/91311924-9d914f00-e7ee-11ea-8042-0971cbbdf02a.png)
![image](https://user-images.githubusercontent.com/21363169/91311974-a97d1100-e7ee-11ea-9bd2-5646b72d6509.png)

#### content cost
![image](https://user-images.githubusercontent.com/21363169/91313086-f01f3b00-e7ef-11ea-9902-0da2aca138e3.png)

#### style cost
![image](https://user-images.githubusercontent.com/21363169/91313053-e39ae280-e7ef-11ea-8b54-61d725c5e648.png)
![image](https://user-images.githubusercontent.com/21363169/91312952-c5cd7d80-e7ef-11ea-9dbe-58e9da3319df.png)
![image](https://user-images.githubusercontent.com/21363169/91313014-d7af2080-e7ef-11ea-87c5-8fa10b3ca9e8.png)

#### Generated image correlated with content image
Now, we initialize the "generated" image as a noisy image created from the content_image.
- The generated image is slightly correlated with the content image.
- By initializing the pixels of the generated image to be mostly noise but slightly correlated with the content image, this will help the content of the "generated" image more rapidly match the content of the "content" image.
![image](https://user-images.githubusercontent.com/21363169/91313297-2fe62280-e7f0-11ea-89f1-b1b9687ffcee.png)

### Programming Assignment: Face Recognition
![image](https://user-images.githubusercontent.com/21363169/91312235-f82aab00-e7ee-11ea-8fd4-8753567414d6.png)

### Kaggle 실습
- https://www.kaggle.com/wonhongyoo/artistic-neural-style-transfer-using-pytorch?scriptVersionId=41463482

### tensorflow turotials / generative / neural style transfer
- https://www.tensorflow.org/tutorials/generative/style_transfer
